tests/test_s3_analytics_provider.py

```diff
@@ -0,0 +1,281 @@
+            self._db,
+            ExternalIntegrationLink,
+            external_integration_id=self._analytics_integration.id,
+            purpose=ExternalIntegrationLink.ANALYTICS,
+        )
+
+        # Act, Assert
+        with pytest.raises(CannotLoadConfiguration):
+            self._analytics_provider.collect_event(
+                self._default_library,
+                None,
+                CirculationEvent.NEW_PATRON,
+                datetime.datetime.utcnow(),
+            )
+
+    def test_exception_is_raised_when_there_is_no_analytics_bucket(self):
+        # Arrange
+        # Create a storage service
+        storage_integration, _ = create(
+            self._db,
+            ExternalIntegration,
+            goal=ExternalIntegration.STORAGE_GOAL,
+            protocol=ExternalIntegration.S3,
+        )
+
+        # Create an external integration link to the storage service
+        create(
+            self._db,
+            ExternalIntegrationLink,
+            external_integration_id=self._analytics_integration.id,
+            other_integration_id=storage_integration.id,
+            purpose=ExternalIntegrationLink.ANALYTICS,
+        )
+
+        # Act, Assert
+        with pytest.raises(CannotLoadConfiguration):
+            self._analytics_provider.collect_event(
+                self._default_library,
+                None,
+                CirculationEvent.NEW_PATRON,
+                datetime.datetime.utcnow(),
+            )
+
+    def test_analytics_data_without_associated_license_pool_is_correctly_stored_in_s3(
+        self,
+    ):
+        # Arrange
+        # Create an S3 Analytics integration
+        analytics_integration, _ = create(
+            self._db,
+            ExternalIntegration,
+            goal=ExternalIntegration.ANALYTICS_GOAL,
+            protocol=S3AnalyticsProvider.__module__,
+        )
+        # Create an S3 Analytics provider
+        provider = S3AnalyticsProvider(analytics_integration, self._default_library)
+
+        # Create an S3 storage service
+        storage_integration, _ = create(
+            self._db,
+            ExternalIntegration,
+            goal=ExternalIntegration.STORAGE_GOAL,
+            protocol=ExternalIntegration.S3,
+        )
+        # Set up a bucket name used for storing analytics data
+        storage_integration.setting(
+            S3UploaderConfiguration.ANALYTICS_BUCKET_KEY
+        ).value = "analytics"
+
+        # Create a link to the S3 storage service
+        create(
+            self._db,
+            ExternalIntegrationLink,
+            external_integration_id=analytics_integration.id,
+            other_integration_id=storage_integration.id,
+            purpose=ExternalIntegrationLink.ANALYTICS,
+        )
+
+        # Set up a mock instead of real S3Uploader class acting as the S3 storage service
+        s3_uploader = create_autospec(spec=S3Uploader)
+        MirrorUploader.implementation = MagicMock(return_value=s3_uploader)
```

Comment:
A common problem that may occur due to not using either the `mock.patch` context manager or decorator is that we've permanently patched the `MirrorUploader.implementation` for all subsequent tests. 
Not only for this test class but for any tests running after this point, this may not be causing problems now but may be a "gotcha" down the line.

Reply:
I'll fix - good catch.